runtime-sdk/src/storage/confidential.rs

```diff
@@ -0,0 +1,400 @@
+use std::convert::TryInto as _;
+
+use anyhow;
+use slog::error;
+use thiserror::Error;
+use zeroize::Zeroize;
+
+pub use crate::core::common::crypto::mrae::deoxysii::KEY_SIZE;
+use crate::{
+    core::{
+        common::crypto::{
+            hash::Hash,
+            mrae::deoxysii::{self, NONCE_SIZE},
+        },
+        storage::mkvs,
+    },
+    storage::Store,
+};
+
+type Nonce = [u8; NONCE_SIZE];
+
+pub struct EncryptionKey([u8; KEY_SIZE]);
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("corrupt key")]
+    CorruptKey,
+
+    #[error("corrupt value")]
+    CorruptValue,
+
+    #[error("decryption failure: {0}")]
+    DecryptionFailure(anyhow::Error),
+}
+
+/// A key-value store that encrypts all content with DeoxysII.
+pub struct ConfidentialStore<S: Store> {
+    inner: S,
+    deoxys: deoxysii::DeoxysII,
+}
+
+impl<S: Store> ConfidentialStore<S> {
+    /// Create a new confidential store with the given keypair.
+    pub fn new_with_key(inner: S, key: [u8; KEY_SIZE]) -> Self {
+        let actual_key = zeroize::Zeroizing::new(key);
+        ConfidentialStore {
+            inner,
+            deoxys: deoxysii::DeoxysII::new(&actual_key),
+        }
+    }
+
+    fn pack_key(&self, enc_key: &[u8], nonce: &Nonce) -> Vec<u8> {
+        let mut ret = Vec::with_capacity(nonce.len() + enc_key.len());
+        ret.extend_from_slice(nonce);
+        ret.extend_from_slice(enc_key);
+        ret
+    }
+
+    fn make_key(&self, plain_key: &[u8]) -> (Nonce, Vec<u8>) {
+        let mut nonce = [0u8; NONCE_SIZE];
+        let plain_hash = Hash::digest_bytes(plain_key);
```

Comment:
This could probably leak some information about `plain_key`. Some better options:
* Use a value derived from something like the current `round || tx index || internal counter`, where the internal counter is incremented every time you need a new nonce within the same instance. You can get the first two from context and pass them to the constructor.
* Derive a "nonce key" from `key` and a KDF when initializing the instance in `new_with_key`. Use `H(nonce_key || plain_key)` here. The downside is that this will result in the same nonce for the same `plain_key`.

@Yawning any thoughts?

Reply:
Ooops, the first scheme wouldn't work for keys, only for values.

Reply:
By virtue of it being the key rather than the value, leaking "keys are identical" is somewhat hard, but the keypair used to initialize the store is presumably providing domain separation (at the contract level?), so I don't think that's a big deal...